### PR TITLE
Custom Attributes

### LINKED ISSUE #117: Custom resource attributes
**Is your feature request related to a problem? Please describe.**
NetSuite (and other ledgers) support customer fields/attributes on resources.  Right now, you must specify each attribute per resource (and in operation validations).

**Thoughts/Ideas**

1. `ResourceAttributeSet`
Right now we store all attributes in a `ResourceAttributeSet` object.  We could modify resources to have an `.attributes` instance method instead of attributes being directly accessible.  Then we can have that set accept any key/value pair.

2. Add a `custom_attributes: {}` keyword to resource initializer

3. Add a way to specify custom attributes when calling `LedgerSync.register_resource` (or somewhere when instantiating LedgerSync)

This would require devs to know and specify their custom attributes.  NetSuite does provide a dynamic schema/metadata through their API.  I don't know that this holds for other ledgers that offer this level of customization.  In theory, they may just return custom attributes as part of the record's response body and we may not have a formal way to detect.  So perhaps being intentional and having devs define these attributes makes sense.

4. Add raw body to resource

Ideally, resources will remain adaptor independent.  But there is data loss (even now) for all the attributes we aren't defining.  You can access this data through the `Response` object body, but it's pretty inconvenient.  Perhaps we can add an `adaptor_body` attribute to resources or something?

**Appendix**
For all of these we would have to determine how to handle operation validations, which right now require that you define a validation for every attribute.

### PR DESCRIPTION
**Pull request by ryanwjackson**
Fix #117 

@mattgmarcus @jozefvaclavik this is ready for review.  I added examples to the README, but please let me know if that's not clear.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -96,18 +94,67 @@ Resources have defined attributes.  Attributes are explicitly defined.  An error
...
 
+#### Custom Attributes
+
+Some ledgers (e.g. NetSuite) allow for custom attributes (or "fields"), which can vary by account and user.  To allow for custom attributes, you can create new resources, ledger serializers (see below), and validation contracts (see below).  Assuming your ledger supports the string attribute `foo` for customers, you could do the following:
+
+```ruby
+class CustomCustomer < LedgerSync::Customer
+  attribute :foo, type: LedgerSync::Type::String
+end
+```
```
Comment (mattgmarcus) [inline README.md]: Can you put an example of setting this like this for documentation
```
klass = Class.new(LedgerSync::Customer)
klass.attribute :foo, type: LedgerSync::Type::String

```

File: README.md [hunk 2]
```diff
@@ -96,18 +94,67 @@ Resources have defined attributes.  Attributes are explicitly defined.  An error
...
+Operations depend on `LedgerSync::Adaptor::LedgerSerializer`s to serialize and deserialize objects.  Most resources have a default serializer per adaptor which may acts as both a serializer and deserializer.  You can provide custom serializers, which is necessary when working with custom attributes.  For example, given the following:
+- `custom_resource` that is a `LedgerSync::Customer` (see above)
+- the attribute `foo` is used in both the request and response bodies
+- using the `NetSuite` adaptor
+
+you could implement custom serializers using the following code:
+
+```ruby
+# Extend the existing one or use
+# LedgerSync::Adaptors::NetSuite::LedgerSerializer or, more generically,
```
Comment (mattgmarcus) [inline README.md]: Should `LedgerSync::Adaptors::NetSuite::LedgerSerializer` be `LedgerSync::Adaptors::NetSuite::Customer::LedgerSerializer` ?

File: lib/ledger_sync/adaptors/ledger_serializer.rb [hunk 3]
```diff
@@ -12,9 +12,8 @@ class LedgerSerializer
 
       attr_reader :resource
 
-      def initialize(ensure_inferred_resource_class: true, resource:)
+      def initialize(resource:)
```
Comment (mattgmarcus) [inline lib/ledger_sync/adaptors/ledger_serializer.rb]: We had talked about maybe altering this in such a way that this interface doesn't break going forward. Up to you

File: lib/ledger_sync/adaptors/operation.rb [hunk 4]
```diff
@@ -46,16 +48,34 @@ def self.included(base)
...
-          raise 'Missing resource' if resource.nil?
-
-          raise "#{resource.class.name} is not a valid resource type.  Expected #{self.class.resource_klass.name}" unless resource.is_a?(self.class.resource_klass)
+        def initialize(
+          adaptor:,
+          ledger_deserializer_class: nil,
+          ledger_serializer_class: nil,
+          resource:,
+          validation_contract: nil
+        )
```
Comment (mattgmarcus) [inline lib/ledger_sync/adaptors/operation.rb]: Same point here about interface

File: lib/ledger_sync/adaptors/operation.rb [hunk 5]
```diff
@@ -46,16 +48,34 @@ def self.included(base)
...
+          self.class.raise_if_unexpected_class(expected: self.class.resource_klass, given: resource.class)
 
           @adaptor = adaptor
+
+          unless ledger_deserializer_class.nil?
```
Comment (mattgmarcus) [inline lib/ledger_sync/adaptors/operation.rb]: `if x.present?` seems easier to me than reading `unless x.nil?` IMO

File: lib/ledger_sync/error.rb [hunk 6]
```diff
@@ -1,10 +1,43 @@
+# frozen_string_literal: true
+
 module LedgerSync
   class Error < StandardError
+    module HelpersMixin
+      module ClassMethods
+        def raise_if_unexpected_class(expected:, given:)
+          expected = [expected] unless expected.is_a?(Array)
```
Comment (mattgmarcus) [inline lib/ledger_sync/error.rb]: `expected = Array.wrap(expected)` I think basically does this, if you're interested in using that. Doesn't matter though

File: README.md (hunk 2 above)
Comment (ryanwjackson) [reply README.md]: Not necessarily.  The inheritance looks like this:

` LedgerSync::Adaptors::NetSuite::Customer::LedgerSerializer <  LedgerSync::Adaptors::NetSuite::LedgerSerializer <  LedgerSync::Adaptors::LedgerSerializer`

It really just depends on what you're looking to accomplish.

I updated the readme on this to hopefully be a bit more clear.

File: lib/ledger_sync/adaptors/operation.rb (hunk 5 above)
Comment (ryanwjackson) [reply lib/ledger_sync/adaptors/operation.rb]: If it's `nil` it defaults it (hence the function call two lines below).  If it's present, then it does a type check

**Comment by ryanwjackson** [comment]
@mattgmarcus @SeanBolt Please merge on approval.

**Pull request merged**